src/timemachine.c

```diff
@@ -0,0 +1,613 @@
+static void TimeMachineDeInitCtx(OutputCtx *);
+
+void TmModuleTimeMachineRegister(void)
+{
+    tmm_modules[TMM_TIMEMACHINE].name = MODULE_NAME;
+    tmm_modules[TMM_TIMEMACHINE].ThreadInit = TimeMachineInit;
+    tmm_modules[TMM_TIMEMACHINE].Func = TimeMachineProcess;
+    tmm_modules[TMM_TIMEMACHINE].ThreadDeinit = TimeMachineDeinit;
+    tmm_modules[TMM_TIMEMACHINE].RegisterTests = NULL;
+
+    OutputRegisterModule(MODULE_NAME, "timemachine", TimeMachineInitCtx);
+
+    SC_ATOMIC_INIT(thread_cnt);
+    return;
+}
+
+/**
+ * \brief TimeMachine main logging function
+ *
+ * \param t threadvar
+ * \param p packet
+ * \param data thread module specific data
+ * \param pq pre-packet-queue
+ * \param postpq post-packet-queue
+ *
+ * \retval TM_ECODE_OK on succes
+ * \retval TM_ECODE_FAILED on serious error
+ */
+static TmEcode TimeMachineProcess(ThreadVars *t, Packet *p, void *thread_data, PacketQueue *pq,
+                 PacketQueue *postpq)
+{    
+    TimeMachineFlow *flow;
+    TimeMachineHeap *heap;
+    TimeMachineMemPool *mem_pool;
+    TimeMachinePacket *packet;
+    struct pcap_pkthdr pkthdr;
+    struct timeval current_time;
+    
+    TimeGet(&current_time);
+    
+    TimeMachineThreadData *td = (TimeMachineThreadData *)thread_data;
+    TimeMachineData *tm = td->tm_data;
+    
+    /* make sure packet has real data */
+    if (GET_PKT_PAYLOAD_LEN(p) == 0) {
+        return TM_ECODE_OK;
+    }
+    
+    /* check whether a corresponding flow already exists */
+    if (!(flow = TimeMachineFlowLookup(td->flows, p))) {
+        flow = TimeMachineFlowNew(td->flows, p);       
+        td->flow_count++;
+        flow->td = td;
+    }
+       
+    /* if this packet has an alert but flow doesn't have a output, create one */
+    if (p->alerts.cnt > 0 && flow->output == NULL) {
+        flow->output = TimeMachineOutputNew(flow);
+        TAILQ_INSERT_TAIL(&td->outputs, flow->output, next);
+        td->output_count += 1;
+    }
+     
+     /* create the packet header info */
+    pkthdr.ts.tv_sec = p->ts.tv_sec;
+    pkthdr.ts.tv_usec = p->ts.tv_usec;
+    pkthdr.caplen = GET_PKT_LEN(p);
+    pkthdr.len = GET_PKT_LEN(p);
+                                         
+    /* dump the traffic if this has already been marked as a output */
+    if (flow->output != NULL) {
+        pcap_dump((u_char*)flow->output->output_file, &pkthdr, GET_PKT_DATA(p));
+        fflush(flow->output->output_file);
+        memcpy(&flow->output->updated, &current_time, sizeof(struct timeval));
+        
+        /* make this the most recently accessed output */
+        TAILQ_REMOVE(&td->outputs, flow->output, next);
+        TAILQ_INSERT_TAIL(&td->outputs,flow->output, next);
+        
+        return TM_ECODE_OK;
+    } 
+    
+    /* find the queue the packet should fall in */
+    TAILQ_FOREACH(heap, &td->heaps, next) {
+        if (pkthdr.caplen <= heap->conf->max_packet_size) {
+            break;
+        }
+    }
+   
```

Comment:
If pkthdr.caplen is > 1500, heap will be undefined here causing a segv at line 160.